src/preprocessing/passes/synth_rew_rules.cpp

```diff
@@ -18,143 +18,436 @@
-        if (cur.getAttribute(SynthRrComputedAttribute()))
+        Trace("srs-input-debug") << "...preprocess " << cur << std::endl;
+        visited[cur] = false;
+        Kind k = cur.getKind();
+        bool isQuant = k == FORALL || k == EXISTS || k == LAMBDA || k == CHOICE;
+        // we recurse on this node if it is not a quantified formula
+        if (!isQuant)
         {
-          Trace("synth-rr-pass-debug")
-              << "...already processed " << cur << std::endl;
+          visit.push_back(cur);
+          for (const Node& cc : cur)
+          {
+            visit.push_back(cc);
+          }
         }
-        else if (it == visited.end())
+      }
+      else if (!it->second)
+      {
+        Trace("srs-input-debug") << "...postprocess " << cur << std::endl;
+        // check if all of the children are valid
+        // this ensures we do not register terms that have e.g. quantified
+        // formulas as subterms
+        bool childrenValid = true;
+        for (const Node& cc : cur)
         {
-          Trace("synth-rr-pass-debug") << "...preprocess " << cur << std::endl;
-          visited[cur] = false;
-          Kind k = cur.getKind();
-          bool isQuant = k == kind::FORALL || k == kind::EXISTS
-                         || k == kind::LAMBDA || k == kind::CHOICE;
-          // we recurse on this node if it is not a quantified formula
-          if (!isQuant)
+          Assert(visited.find(cc) != visited.end());
+          if (!visited[cc])
           {
-            visit.push_back(cur);
-            for (const Node& cc : cur)
-            {
-              visit.push_back(cc);
-            }
+            childrenValid = false;
           }
         }
-        else if (!it->second)
+        if (childrenValid)
         {
-          Trace("synth-rr-pass-debug") << "...postprocess " << cur << std::endl;
-          // check if all of the children are valid
-          // this ensures we do not register terms that have e.g. quantified
-          // formulas as subterms
-          bool childrenValid = true;
-          for (const Node& cc : cur)
+          Trace("srs-input-debug") << "...children are valid" << std::endl;
+          Trace("srs-input-debug") << "Add term " << cur << std::endl;
+          if (cur.isVar())
           {
-            Assert(visited.find(cc) != visited.end());
-            if (!visited[cc])
-            {
-              childrenValid = false;
-            }
+            vars.push_back(cur);
           }
-          if (childrenValid)
+          // register type information
+          TypeNode tn = cur.getType();
+          if (tvars.find(tn) == tvars.end())
           {
-            Trace("synth-rr-pass-debug")
-                << "...children are valid, check rewrites..." << std::endl;
-            if (r == 0)
+            // Only make one Boolean variable unless option is set. This ensures
+            // we do not compute purely Boolean rewrites by default.
```

Comment:
what is the motivation for this?

Reply:
We don't want any input to generate Boolean rewrites, e.g. (or A A) -> A, (and A A) -> A, etc.

Reply:
Again, this should be said somewhere, probably in the header of the class. It's a simple thing that is clear once said but a bit mysterious if you are not used to it.